lib/elixir/lib/kernel.ex

```diff
@@ -3543,17 +3543,28 @@ defmodule Kernel do
       [] when not in_module? ->
         false
 
-      [head | tail] = list when not in_module? ->
-        in_var(in_module?, left, &in_list(&1, head, tail, expand, list, in_module?))
+      [] ->
+        quote(do: :lists.member(unquote(left), []))
+
+      [head | tail] when not in_module? ->
+        in_var(
+          in_module?,
+          left,
+          &in_list(&1, head, tail, expand, right, in_module?, true, [])
+        )
 
       [_ | _] = list when in_module? ->
```

Comment:
We don't need to check the guard again, so we could do:

```suggestion
      [_ | _] = list ->
```

But possibly better is to move the clause above down and keep the guard in this one.